index.src.html

```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
+
+dictionary ReportingObserverOptions {
+  sequence&lt;DOMString> types;
+  boolean futureOnly = false;
+};
+  </pre>
+
+  Issue: How do we specify which reports are observable?  Network-layer reports
+  like CSP reports will not be.  Maybe there's something about the settings
+  object?
+
+  A <dfn id=dom-report interface>Report</dfn> is the application exposed
+  representation of a <a>report</a>. <dfn attribute for="Report">type</dfn>
+  returns [=report/type=], <dfn attribute for="Report">url</dfn> returns
+  [=report/url=], and <dfn attribute for="Report">body</dfn> returns
+  [=report/body=].
+
+  Each {{ReportingObserver}} object has these associated concepts:
+    - A <dfn for=ReportingObserver>callback</dfn> function set on creation.
+    - A {{ReportingObserverOptions}} dictionary called
+      <dfn for=ReportingObserver>options</dfn>.
+    - A list of {{Report}} objects called the <dfn for=ReportingObserver>report
+      list</dfn>, which is initially empty.
+
+  The <dfn constructor for=ReportingObserver><code>
+  ReportingObserver(|callback|, |options|)</code></dfn> constructor, when invoked, must run
+  these steps:
+
+  1. Create a new {{ReportingObserver}} object |observer|.
+
+  2. Set |observer|'s <a>callback</a> to |callback|.
+
+  3. Set |observer|'s <a>options</a> to |options|.
+
+  4. Return |observer|.
+
+  The <dfn method for=ReportingObserver><code>observe()</code></dfn>
+  method, when invoked, must run these steps:
+
+  1. Let |settings| be the <a>environment settings object</a> associated with
+     the <a>context object</a>.
+
+  2. Append the <a>context object</a> to the <a>registered reporting observer
+     list</a> of its associated <a>environment settings object</a>.
+
+  3. If the <a>context object</a>'s {{futureOnly}} <a>option</a> is true, return.
+
+  4. Set <a>context object</a>'s {{futureOnly}} <a>option</a> to true.
+
+  5. For each |report| in the <a>report buffer</a> associated with |settings|,
+     execute [[#add-report]] with |report| and the <a>context object</a>.
+
+  The <dfn method for=ReportingObserver><code>disconnect()</code></dfn> method,
+  when invoked, must run these steps:
+
+  1. If the <a>context object</a> is not <a>registered</a>, return.
+
+  2. Remove the <a>context object</a> from the <a>registered reporting observer
+     list</a> of its associated <a>environment settings object</a>.
+
+  <h3 id="notify-observers" algorithm>
+    Notify reporting observers on |settings| with |report|
+  </h3>
+
+  This algorithm makes |report|'s contents available to any <a>registered</a>
+  <a>reporting observers</a> on the provided <a>environment settings object</a>
+  |settings|.
+
+  1. For each {{ReportingObserver}} |observer| <a>registered</a> with
+     |settings|, execute [[#add-report]] on |report| and |observer|.
+
+  2. Add |report| to the <a>report buffer</a> associated with |settings|.
+
+  3. If the <a>report buffer</a> now contains more than 100 reports, remove the
```

Comment:
This implies there is a global 100 reports limit? 

What is the "beginning"? We should spell it out.

Reply:
Yes, there is a 100 report buffer limit.

The "beginning" of the list is used here: https://infra.spec.whatwg.org/#lists, so I think it is unambiguous. If it does need specifying though, then it should probably be specified in infra rather than in this spec.